### ISSUE TITLE
Istio-Pilot on k8s re-created (deleted) rules CRDs during scale-up to 200+ instances

### ISSUE THREAD

**Issue by alex-koturanov** [istio/istio#6977]
**Describe the bug**
Istio-0.7.1 with Pilot, Mixer and Ingress installed as DaemonSet on Kubernetes 1.9.3 with 3 masters deleted and re-created some of the Istio CRDs during cluster scale up from 100 to 250 nodes. **All traffic and egress rules were deleted as a result.** Kubernetes API logs show that the CRDs were deleted by Istio Pilot component after unsuccessful attempt to update a config map. The issue was re-created by scaling the cluster back to 100 nodes and then scaling up to 250 nodes.

**Expected behavior**
No rules loss (traffic, egress, etc).

**Steps to reproduce the bug**
Istio-0.7.1 installation with no Auth. Scale up Pilot deployment to 250 replicas. Run `kubectl get crd` and check the timestamps.

**Version**
What version of istio and Kubernetes are you using? Use `istioctl version` and `kubectl version`
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T12:22:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
````

**Is Istio Auth enabled or not?**
Did you install the stable istio.yaml, istio-auth.yaml.... or if using the Helm chart please provide full command line input.
```
istio.yaml from 0.7.1 release (no auth)
```

**Environment**
Which environment, cloud vendor, OS, etc are you using?
AWS KOPS 1.9

**List of re-created CRDs**
```
List of Istio CRDs that were re-created during Pilot scale -up

kubectl get crd | grep istio | grep 2h

destinationpolicies.config.istio.io                       2h
egressrules.config.istio.io                               2h
enduserauthenticationpolicyspecbindings.config.istio.io   2h
enduserauthenticationpolicyspecs.config.istio.io          2h
httpapispecbindings.config.istio.io                       2h
httpapispecs.config.istio.io                              2h
quotaspecbindings.config.istio.io                         2h
quotaspecs.config.istio.io                                2h
routerules.config.istio.io                                2h
servicerolebindings.config.istio.io                       2h
serviceroles.config.istio.io                              2h
```

**Timestamp of a re-created CRD**
```
$ kubectl get crd egressrules.config.istio.io -o yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: 2018-06-28T23:01:18Z
  generation: 1
  name: egressrules.config.istio.io
  resourceVersion: "134596540"
  selfLink: /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io
  uid: 2aa8558d-7b27-11e8-b7f1-0e5af0fd0eb6
spec:
  group: config.istio.io
  names:
    kind: EgressRule
    listKind: EgressRuleList
    plural: egressrules
    singular: egressrule
  scope: Namespaced
  version: v1alpha2
status:
  acceptedNames:
    kind: EgressRule
    listKind: EgressRuleList
    plural: egressrules
    singular: egressrule
  conditions:
  - lastTransitionTime: 2018-06-28T23:01:18Z
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: 2018-06-28T23:01:18Z
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
```

**Kubernetes Master API logs arount the time CRDs were re-created, note PUT/DELETE operations**

```
MASTER-1

sudo grep "egressrules" /var/log/20180628-copy/kube-apiserver.log.2|less

I0628 23:01:04.696177       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (445.531�s) 200 [[Go-http-client/
2.0] 172.20.123.99:57266]
I0628 23:01:04.697117       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587363 labels= fields= timeout=7m6
s
I0628 23:01:11.088070       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=76910922&timeoutSeconds=354: (5m54.000455304s) 0 [[Go-http-client/2.0] 172.20.83.53:46862]
I0628 23:01:11.089575       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587198 labels= fields= timeout=9m52s
I0628 23:01:11.089898       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=134587198&timeoutSeconds=592: (546.95�s) 0 [[Go-http-client/2.0] 172.20.83.53:46862]
I0628 23:01:12.091830       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (395.231�s) 200 [[Go-http-client/2.0] 172.20.83.53:46862]
I0628 23:01:12.093445       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587363 labels= fields= timeout=8m28s
I0628 23:01:12.948684       7 trace.go:76] Trace[1107715927]: "Update /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status" (started: 2018-06-28 23:01:12.300980159 +0000 UTC m=+8753027.570282938) (total time: 647.659812ms):
I0628 23:01:12.948819       7 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (647.874591ms) 409 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:17578]
E0628 23:01:12.949246       7 crd_finalizer.go:275] egressrules.config.istio.io failed with: Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io "egressrules.config.istio.io": the object has been modified; please apply your changes to the latest version and try again
I0628 23:01:13.158828       7 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (203.80913ms) 409 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:17578]
E0628 23:01:13.159179       7 crd_finalizer.go:275] egressrules.config.istio.io failed with: Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io "egressrules.config.istio.io": the object has been modified; please apply your changes to the latest version and try again
I0628 23:01:26.181328       7 wrap.go:42] GET /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (120.999141ms) 200 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.118.193:32776]
I0628 23:01:26.878672       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=76910922&timeoutSeconds=448: (7m28.000366203s) 0 [[Go-http-client/2.0] 172.20.75.219:52282]
```

```
MASTER-2

sudo grep "egressrules" /var/log/20180628-copy/kube-apiserver.log.2|less

I0628 23:00:54.080712       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=134587198&timeoutSeconds=306: (509.644�s
) 0 [[Go-http-client/2.0] 172.20.85.72:56518]
I0628 23:00:54.613926       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (236.464�s) 200 [[Go-http-client/
2.0] 172.20.64.48:48322]
I0628 23:00:54.615582       6 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587409 labels= fields= timeout=6m5
5s
I0628 23:00:55.082117       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (236.91�s) 200 [[Go-http-client/2
.0] 172.20.85.72:56518]
I0628 23:00:55.083269       6 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587409 labels= fields= timeout=7m6
s
I0628 23:00:55.503252       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=76910922&timeoutSeconds=383: (6m23.00050
894s) 0 [[Go-http-client/2.0] 172.20.49.14:60986]
I0628 23:00:55.504546       6 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587198 labels= fields= timeout=9m2
9s
I0628 23:00:55.504923       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=134587198&timeoutSeconds=569: (668.739�s
) 0 [[Go-http-client/2.0] 172.20.49.14:60986]
I0628 23:00:56.505954       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (323.082�s) 200 [[Go-http-client/
2.0] 172.20.49.14:60986]
I0628 23:00:56.506972       6 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587409 labels= fields= timeout=8m2
4s
I0628 23:01:12.948012       6 trace.go:76] Trace[1779088624]: "Update /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.i
stio.io/status" (started: 2018-06-28 23:01:12.30368915 +0000 UTC m=+8753380.246828918) (total time: 644.299769ms):
I0628 23:01:12.948079       6 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (644.497
879ms) 409 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:15242]
E0628 23:01:12.948328       6 crd_finalizer.go:275] egressrules.config.istio.io failed with: Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io "egressrules.config.istio.io": the object has been modified; please apply your changes to the latest version and try again
I0628 23:01:13.158202       6 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (204.315016ms) 409 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:15242]
E0628 23:01:13.158726       6 crd_finalizer.go:275] egressrules.config.istio.io failed with: Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io "egressrules.config.istio.io": the object has been modified; please apply your changes to the latest version and try again

I0628 23:01:14.778770       6 wrap.go:42] DELETE /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (198.953691ms) 404 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.90.107:46548]

I0628 23:01:22.260417       6 wrap.go:42] GET /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (210.718784ms) 200 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.42.41:46670]
I0628 23:01:29.355767       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (209.588755ms) 200 [[Go-http-client/2.0] 172.20.42.41:46670]
I0628 23:01:29.357754       6 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134596869 labels= fields= timeout=6m0s
I0628 23:01:39.181565       6 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=134587409&timeoutSeconds=306: (5m6.000463124s) 0 [[Go-http-client/2.0] 172.20.96.217:43340]
```

```
MASTER-3

sudo grep "egressrules" /var/log/20180628-copy/kube-apiserver.log.2|less

I0628 23:00:51.800516       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587837 labels= fields= timeout=6m4
s
I0628 23:00:53.523292       7 wrap.go:42] GET /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (196.84581ms) 2
00 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.85.30:53376]
I0628 23:01:01.433891       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (358.176�s) 200 [[Go-http-client/
2.0] 172.20.85.30:53376]
I0628 23:01:01.442437       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587837 labels= fields= timeout=8m2
3s
I0628 23:01:12.268106       7 wrap.go:42] DELETE /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (321.563755ms) 200 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.111.129:50840]
I0628 23:01:12.537688       7 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (268.784491ms) 200 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:51750]
I0628 23:01:13.004012       7 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (239.852002ms) 200 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:51750]
I0628 23:01:13.463846       7 wrap.go:42] DELETE /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (459.209201ms) 200 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:51750]
I0628 23:01:18.884318       7 wrap.go:42] PUT /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io/status: (239.531237ms) 200 [[kube-apiserver/v1.9.3 (linux/amd64) kubernetes/d283541] 127.0.0.1:51750]
I0628 23:01:25.528542       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=76910922&timeoutSeconds=427: (7m7.058025419s) 0 [[Go-http-client/2.0] 172.20.117.184:47034]
I0628 23:01:25.530555       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134587198 labels= fields= timeout=6m23s
I0628 23:01:26.025144       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/watch/egressrules?resourceVersion=134587198&timeoutSeconds=383: (495.068017ms) 0 [[Go-http-client/2.0] 172.20.117.184:47034]
I0628 23:01:27.026852       7 wrap.go:42] GET /apis/config.istio.io/v1alpha2/egressrules?limit=500&resourceVersion=0: (328.583�s) 200 [[Go-http-client/2.0] 172.20.117.184:47034]
I0628 23:01:27.030383       7 get.go:238] Starting watch for /apis/config.istio.io/v1alpha2/watch/egressrules, rv=134596750 labels= fields= timeout=7m6s
I0628 23:01:30.855270       7 wrap.go:42] GET /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/egressrules.config.istio.io: (118.099177ms) 200 [[pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format] 172.20.75.25:53190]
```

**Comment by kdillane** [comment]
It looks like the line of code causing this problem could possibly be traced to these lines: https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L206
https://github.com/istio/istio/blob/release-1.0/pilot/pkg/bootstrap/server.go#L517

If the API server is unreachable, say due to a scaling event, you could get back an error causing the resources to be recreated.  If polling fails, the resources can be deleted.  If there are a number of pilots coming up together, this can cause the above result.

**Comment by kdillane** [comment]
We were bitten by this problem again.  The underlying issue is likely that our Kubernetes cluster has become unhealthy causing intermittent authentication errors to the api server.  Regardless, the problem is not that Kubernetes is unhealthy.  Pilot, in what I assume is an attempt to heal the cluster, removes the CRDs which effectively wipes all the rules from the cluster.

**Comment by kdillane** [comment]
The code that actually deletes resources: https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L288

**Comment by kyessenov** [comment]
This code is from pre-HA and pre-helm dates. I think it's fine to remove
the logic around deleting CRD entirely.

**Comment by kdillane** [comment]
When do you expect to be able to push out a fix for this issue?  Will this be available as a patch release given it's impact?

**Comment by louiscryan** [comment]
@costinm @rshriram can one of your triage this for inclusion in 1.0.2 ?

**Comment by kdillane** [comment]
Any update on triaging this issue?

**Comment by kdillane** [comment]
It looks like Istio 1.0.2 has been released.  Any updates?  I'm going to provide some additional details about what I think can help to reproduce the issue.

**Comment by kdillane** [comment]
You can see for Istio Pilot Deployment for a (mostly) stock Istio install, the Istio Pilot has a "Discovery" container from which this is a snippet of the k8s yaml:
            "name": "discovery",
            "image": "docker.io/istio/pilot:0.7.1",
            "args": [
              "discovery",
              "--admission-service",
              "istio-pilot"
            ],

**Comment by kdillane** [comment]
When the discovery server is started based on those arguments, the code up through release-1.0 executes the initConfigController method: https://github.com/istio/istio/blob/release-1.0/pilot/pkg/bootstrap/server.go#L205

**Comment by kdillane** [comment]
The Kube Config Controller calls out to RegisterResources on the CRD config client: https://github.com/istio/istio/blob/release-1.0/pilot/pkg/bootstrap/server.go#L517.  It is in this method where the bug exists.

**Comment by kdillane** [comment]
If for any reason, the state of the CRDs in Kubernetes cannot be retrieved, registerResources will avoid skipping the creation of CRDs:  https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L209.    While there is an error case during CRD registration which will cause the method to return, it is only in the case where the CRDs are detected as existing:  https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L256.

**Comment by kdillane** [comment]
In the case where Istio attempts to create CRDs, but does not detect an "Exists" error, during the polling of CRDs, it is likely for Istio to begin polling for CRDs and fail while reading:  https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L265.

**Comment by kdillane** [comment]
In the case that polling for resources fails, Istio  will attempt to deregister all the CRDs:  https://github.com/istio/istio/blob/release-1.0/pilot/pkg/config/kube/crd/client.go#L289

**Comment by kdillane** [comment]
However, unlikely these sequence of events might be, we've observed this failure mode multiple times.

**Comment by kdillane** [comment]
One other point, is in the bug listed above, I've produced a repro case where if only read permissions are revoked from Pilot, then all Istio CRDs are deleted from k8s.  I believe this case is a lot easier to hit than I previously anticipated.

**Comment by kdillane** [comment]
Closed #8691 in favor of https://github.com/istio/istio/pull/8813

**Comment by kdillane** [comment]
PR #8813 has been merged.  Awaiting release of the patch with 1.0.3 and this issue can be closed.

**Comment by louiscryan** [comment]
Assigning for triage

**Comment by mandarjog** [comment]
fixed by #8813
